File: app/src/main/java/org/ole/planet/myplanet/service/AudioRecorderService.java
package org.ole.planet.myplanet.service;

import static org.ole.planet.myplanet.MainApplication.context;

import android.database.Cursor;
import android.media.MediaRecorder;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;

import java.io.File;
import java.util.UUID;

public class AudioRecorderService {
    private String outputFile;
    private MediaRecorder myAudioRecorder;
    private AudioRecordListener audioRecordListener;

    public AudioRecorderService() {
    }

    public void forceStop() {
        if (myAudioRecorder != null) {
            myAudioRecorder.stop();
            myAudioRecorder.release();
            myAudioRecorder = null;
        }
        if (audioRecordListener != null) audioRecordListener.onError("Recording stopped");
    }

    public AudioRecorderService setAudioRecordListener(AudioRecordListener audioRecordListener) {
        this.audioRecordListener = audioRecordListener;
        return this;
    }

    public void startRecording() {
        outputFile = createAudioFile();
        myAudioRecorder = new MediaRecorder();
        myAudioRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
        myAudioRecorder.setOutputFormat(MediaRecorder.OutputFormat.AAC_ADTS);
        myAudioRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);
        myAudioRecorder.setOutputFile(outputFile);
        try {
            myAudioRecorder.prepare();
            myAudioRecorder.start();
            if (audioRecordListener != null) audioRecordListener.onRecordStarted();
        } catch (Exception e) {
            myAudioRecorder = null;
            if (audioRecordListener != null) {
                audioRecordListener.onError(e.getMessage());
            }
        }
    }

    private String createAudioFile() {
        String audioFileName;
        File audioFile;
        int attempt = 0;

        do {
            audioFileName = UUID.randomUUID().toString() + ".aac";
            audioFile = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC), audioFileName);
            attempt++;
        } while (audioFile.exists() && attempt < 100);

        if (attempt >= 100) {
            return null;
        }

        return audioFile.getAbsolutePath();
    }

    private String getFilePathFromUri(Uri uri) {
        String filePath = null;
        String[] projection = {MediaStore.Audio.Media.DATA};
        Cursor cursor = context.getContentResolver().query(uri, projection, null, null, null);
        if (cursor != null && cursor.moveToFirst()) {
            int columnIndex = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DATA);
            filePath = cursor.getString(columnIndex);
            cursor.close();
        }
        return filePath;
    }

    public boolean isRecording() {
        return myAudioRecorder != null;
    }

    public void stopRecording() {
        if (myAudioRecorder != null) {
            myAudioRecorder.stop();
            myAudioRecorder.release();
            myAudioRecorder = null;
            if (audioRecordListener != null) audioRecordListener.onRecordStopped(outputFile);
        }
    }

    public interface AudioRecordListener {
        void onRecordStarted();

        void onRecordStopped(String outputFile);

        void onError(String error);
    }
}

File: app/src/main/java/org/ole/planet/myplanet/service/AudioRecorderService.kt
package org.ole.planet.myplanet.service

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.MediaRecorder
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.Settings
import androidx.activity.result.ActivityResultCaller
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat.shouldShowRequestPermissionRationale
import androidx.core.content.ContextCompat
import org.ole.planet.myplanet.MainApplication
import org.ole.planet.myplanet.MainApplication.Companion.context
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.utilities.Utilities
import java.io.File
import java.util.UUID

class AudioRecorderService {
    private var outputFile: String? = null
    private var myAudioRecorder: MediaRecorder? = null
    private var audioRecordListener: AudioRecordListener? = null
    private var caller: ActivityResultCaller? = null
    private lateinit var permissionLauncher: ActivityResultLauncher<String>

    fun forceStop() {
        myAudioRecorder?.apply {
            stop()
            release()
        }
        myAudioRecorder = null
        audioRecordListener?.onError("Recording stopped")
    }

    fun setAudioRecordListener(audioRecordListener: AudioRecordListener): AudioRecorderService {
        this.audioRecordListener = audioRecordListener
        return this
    }

    @RequiresApi(Build.VERSION_CODES.S)
    fun startRecording() {
        outputFile = createAudioFile()
        myAudioRecorder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }

        myAudioRecorder?.apply {
            setAudioSource(MediaRecorder.AudioSource.MIC)
            setOutputFormat(MediaRecorder.OutputFormat.AAC_ADTS)
            setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            setOutputFile(outputFile)
            try {
                prepare()
                start()
                audioRecordListener?.onRecordStarted()
            } catch (e: Exception) {
                myAudioRecorder = null
                audioRecordListener?.onError(e.message)
            }
        }
    }

    private fun createAudioFile(): String? {
        var audioFileName: String
        var audioFile: File
        var attempt = 0
        do {
            audioFileName = "${UUID.randomUUID()}.aac"
            audioFile = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC), audioFileName)
            attempt++
        } while (audioFile.exists() && attempt < 100)
        if (attempt >= 100) {
            return null
        }
        return audioFile.absolutePath
    }

    fun isRecording(): Boolean {
        return myAudioRecorder != null
    }

    fun stopRecording() {
        myAudioRecorder?.let { recorder ->
            try {
                if (isRecording()) {
                    recorder.stop()
                    recorder.release()
                }
            } catch (e: RuntimeException) {
                MainApplication.handleUncaughtException(e)
            } finally {
                myAudioRecorder = null
                audioRecordListener?.onRecordStopped(outputFile)
            }
        }
    }

    fun setCaller(caller: ActivityResultCaller, context: Context){
        this.caller = caller
        permissionLauncher =
            caller.registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
                if (granted){
                    toggleRecording()
                }
                else {
                    if (!shouldShowRequestPermissionRationale(context as Activity, Manifest.permission.RECORD_AUDIO)) {
                        AlertDialog.Builder(context, R.style.AlertDialogTheme)
                            .setTitle(R.string.permission_required)
                            .setMessage(R.string.microphone_permission_required)
                            .setPositiveButton(R.string.settings) { dialog, _ ->
                                dialog.dismiss()
                                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                                val uri: Uri = Uri.fromParts("package", context.packageName, null)
                                intent.data = uri
                                context.startActivity(intent)
                            }
                            .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
                            .show()
                    } else {
                        Utilities.toast(context, "Microphone permission is required to record audio.")
                    }
                }
            }
    }

    fun onRecordClicked() {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO)
            == PackageManager.PERMISSION_GRANTED) {
            toggleRecording()
        } else {
            permissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
        }
    }

    private fun toggleRecording() {
        if (this.isRecording()) {
            this.stopRecording()
        } else {
            this.startRecording()
        }
    }

    interface AudioRecordListener {
        fun onRecordStarted()
        fun onRecordStopped(outputFile: String?)
        fun onError(error: String?)
    }
}